spot-vaults/contracts/BillBroker.sol

```diff
@@ -0,0 +1,753 @@
+
+/**
+ *  @title BillBroker
+ *
+ *  @notice The `BillBroker` contract (inspired by bill brokers in LombardSt) acts as an intermediary between parties who want to borrow and lend.
+ *
+ *          `BillBroker` LPs deposit perps and dollars as available liquidity into the contract.
+ *          Any user can now sell/buy perps (swap) from the bill broker for dollars, at a "fair" exchange rate determined by the contract.
+ *
+ *          The contract charges a fee for swap operations. The fee is a function of available liquidity held in the contract, and goes to the LPs.
+ *
+ *          The ratio of value of dollar tokens vs perp tokens held by the contract is defined as it's `assetRatio`.
+ *          The owner can define hard limits on the system's assetRatio outside which swapping is disabled.
+ *
+ *          The contract relies on external data sources to price assets.
+ *          If the data is unreliable, swaps are simply halted.
+ *
+ *          Intermediating borrowing:
+ *          Borrowers who want to borrower dollars against their collateral,
+ *          tranche their collateral, mint perps and sell it for dollars to the bill broker.
+ *          When they want to close out their position they can buy back perps from the bill broker
+ *          using dollars, and redeem their tranches for the original collateral.
+ *          The spread is the "interest rate" paid for the loan, which goes to the bill broker LPs
+ *          who take on the risk of holding perps through the duration of the loan.
+ *
+ *          Intermediating lending:
+ *          Lenders can buy perps from the bill broker contract when it's under-priced,
+ *          hold the perp tokens until the market price recovers and sell it back to the bill broker contract.
+ *
+ *
+ */
+contract BillBroker is
+    ERC20BurnableUpgradeable,
+    OwnableUpgradeable,
+    PausableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    //-------------------------------------------------------------------------
+    // Libraries
+
+    // ERC20 operations
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+    using SafeERC20Upgradeable for IPerpetualTranche;
+
+    // Math
+    using MathUpgradeable for uint256;
+    using SafeCastUpgradeable for uint256;
+    using SafeCastUpgradeable for int256;
+    using SignedMathUpgradeable for int256;
+
+    //-------------------------------------------------------------------------
+    // Constants & Immutables
+
+    uint256 private constant DECIMALS = 18;
+    uint256 private constant ONE = (10 ** DECIMALS);
+
+    //-------------------------------------------------------------------------
+    // Storage
+
+    /// @notice The perpetual senior tranche token.
+    IPerpetualTranche public perp;
+
+    /// @notice The USD token.
+    IERC20Upgradeable public usd;
+
+    /// @notice The fixed-point amount of usd tokens equivalent to 1.0$.
+    uint256 public usdUnitAmt;
```

Comment:
The upgradability is what forces us to user the initializer pattern

Reply:
I'm open to hardcoding the usd unit amount, but the number of decimals for perps changes based on the implementation. For SPOT its 9 and if we use other button tokens (like btc/eth) its 18.

Reply:
I was thinking these hardcoded values would be set per deployment still, so each would require a slight code revision. Because once it's deployed, there's no way to change the tokens involved anyway.

Very messy, and not "good practice"--it would be purely a practical tradeoff.

Reply:
I don't think one additional storage read matters as much in the larger scheme of things. The billbroker is not like an amm which does tens of thousands of transactions a day, in practice might do one or two a day ..

Reply:
Ok let's table it for now.

I do think this could become a primary market for Spot, though, which other AMMs are arbed against. After that, it's only a matter of time until primary trade flow gets routed directly to the BillBroker.